File: automcp/adapters/crewai.py
from typing import Any, Callable, Type
from pydantic import BaseModel
import json
import contextlib
import io
import asyncio
import inspect
from automcp.adapters.utils import ensure_serializable

def create_crewai_adapter(
    agent_instance: Any,
    name: str,
    description: str,
    input_schema: Type[BaseModel],
) -> Callable:
        """
        Convert a CrewAI class to an MCP tool.
        
        Args:
            orchestrator_instance: The CrewAI class to convert
            name: The name of the MCP tool
            description: The description of the MCP tool
            input_schema: The Pydantic model class defining the input schema
            
        Returns:
            A callable function that can be used as an MCP tool
        """
        # Get the field names and types from the input schema
        schema_fields = input_schema.model_fields

        # Create the parameter string for the function signature
        params_str = ", ".join(
            f"{field_name}: {field_info.annotation.__name__}"
            for field_name, field_info in schema_fields.items()
        )

        # Create the function body that constructs the input schema
        body_str = f"""def run_agent({params_str}):
            inputs = input_schema({', '.join(f'{name}={name}' for name in schema_fields)})
            with contextlib.redirect_stdout(io.StringIO()):
                result = agent_instance.kickoff(inputs=inputs.model_dump())
            return result.model_dump_json()
        """

        # Create a namespace for the function
        namespace = {
            "input_schema": input_schema,
            "agent_instance": agent_instance,
            "json": json,
            "contextlib": contextlib,
            "io": io,
        }

        # Execute the function definition in the namespace
        exec(body_str, namespace)

        # Get the created function
        run_agent = namespace["run_agent"]

        # Add proper function metadata
        run_agent.__name__ = name
        run_agent.__doc__ = description

        return run_agent

File: src/adapters/crewai.ts
import { ensureSerializable, extractParamsFromArgs, parseInputWithSchema, type SchemaLike } from './utils.js';
import { ToolResult } from './base.js';

interface CrewAIResult {
  model_dump_json(): string;
  model_dump(): Record<string, any>;
}

interface CrewAIAgent {
  kickoff(inputs: Record<string, any>): CrewAIResult;
}

/**
 * Convert a CrewAI class to an MCP tool.
 */
export function createCrewAIAdapter(
  agentInstance: CrewAIAgent,
  name: string,
  description: string,
  inputSchema: SchemaLike
): (...args: any[]) => ToolResult {
  const runAgent = (...args: any[]): ToolResult => {
    const kwargs = extractParamsFromArgs(args, inputSchema);
    const parsedInputs = parseInputWithSchema(inputSchema, kwargs);
    const normalizedInputs = parsedInputs && typeof parsedInputs === 'object'
      ? parsedInputs
      : kwargs;

    // Redirect stdout equivalent (capture console output)
    const originalLog = console.log;
    const originalWarn = console.warn;
    const originalError = console.error;
    const outputBuffer: string[] = [];
    
    // Override console methods to capture output
    console.log = (...args) => outputBuffer.push(args.join(' '));
    console.warn = (...args) => outputBuffer.push(args.join(' '));
    console.error = (...args) => outputBuffer.push(args.join(' '));

    try {
      // Execute CrewAI kickoff
      const result = agentInstance.kickoff({ inputs: normalizedInputs });
      const serializedResult = typeof result?.model_dump === 'function'
        ? result.model_dump()
        : result;
      const outputObj = ensureSerializable(serializedResult);
      return {
        content: [{ type: 'text', text: JSON.stringify(outputObj, null, 2) }],
        structuredContent: outputObj
      };
    } finally {
      // Restore original console methods
      console.log = originalLog;
      console.warn = originalWarn;
      console.error = originalError;
    }
  };
  
  // Add proper function metadata
  Object.defineProperty(runAgent, 'name', { value: name });
  Object.defineProperty(runAgent, 'description', { value: description });
  
  return runAgent;
}

/**
 * Create a typed wrapper for CrewAI adapter with specific input schema
 */
export function createTypedCrewAIAdapter<T extends Record<string, any>>(
  agentInstance: CrewAIAgent,
  name: string,
  description: string,
  inputSchema: SchemaLike
): (input: T) => ToolResult {
  const adapter = createCrewAIAdapter(agentInstance, name, description, inputSchema);

  return (input: T): ToolResult => {
    return adapter(input);
  };
}

/**
 * Mock CrewAI result class for testing
 */
export class MockCrewAIResult implements CrewAIResult {
  private data: Record<string, any>;
  
  constructor(data: Record<string, any>) {
    this.data = ensureSerializable(data) as Record<string, any>;
  }
  
  model_dump_json(): string {
    return JSON.stringify(this.data);
  }
  
  model_dump(): Record<string, any> {
    return this.data;
  }
} 